r1cs/gadgets/std/src/fields/fp2.rs

```diff
@@ -180,131 +228,231 @@ impl<P: Fp2Parameters<Fp = ConstraintF>, ConstraintF: PrimeField + SquareRootFie
+            c0 + c1 *X = (a0 + a1 * X) * (b0 + b1 * X)
+           of a field gadget self =  (a0 + a1 * X) by a constant element (b0 + b1 * X).
+            c0 = a0*b0 + non_residue *a1*b1
+            c1 = a0*b1 + a1*b0
+        Doesn't need any constraints; returns linear combinations of a0, a1.
+        */
+        let (a0, a1) = (&self.c0, &self.c1);
+        let (b0, b1) = (fe.c0, fe.c1);
+        let mut v0 = a0.mul_by_constant(&mut cs.ns(|| "v0"), &b0)?;
+        let beta_v1 = a1.mul_by_constant(&mut cs.ns(|| "non_residue * v1"), &(b1 * &P::NONRESIDUE))?;
+
+        v0.add_in_place(&mut cs.ns(|| "c0 = v0 + non_residue * v1"), &beta_v1)?;
+        let c0 = v0;
+
+        let mut a0b1 = a0.mul_by_constant(&mut cs.ns(|| "a0b1"), &b1)?;
+        let a1b0 = a1.mul_by_constant(&mut cs.ns(|| "a1b0"), &b0)?;
+        a0b1.add_in_place(&mut cs.ns(|| "c1 = a0b1 + a1b0"), &a1b0)?;
+        let c1 = a0b1;
+        Ok(Self::new(c0, c1))
+    }
 
+    /* improves default implementation by 1 constraint.
+    */
+    #[inline]
+    fn square<CS: ConstraintSystem<ConstraintF>>(
+        &self,
+        mut cs: CS,
+    ) -> Result<Self, SynthesisError> {
+        /* From Libsnark/fp2_gadget.tcc
+         Complex squaring for Fp2,
+             result.c0 = A.c0^2 + A.c1^2 * non_residue,
+             result.c1 = 2 * A.c0 * A.c1
+         Using the Karatsuba trick, enforced by two rank 1 constraints,
+             A.c0 * A.c1 = v0,
+             (A.c0 + A.c1) * (A.c0 + non_residue * A.c1) - v0 * (1 + non_residue) = result.c0,
+         and  2*v0 = result.c1 operating directly on LC level.
+
+        Reference:
+        "Multiplication and Squaring on Pairing-Friendly Fields"
+        Devegili, OhEigeartaigh, Scott, Dahab
+        */
+        // constraint 1
         let mut v0 = self.c0.mul(cs.ns(|| "v0"), &self.c1)?;
+        // prepare for constraint 2
         let a0_plus_a1 = self.c0.add(cs.ns(|| "a0 + a1"), &self.c1)?;
-
         let non_residue_c1 = self
             .c1
             .mul_by_constant(cs.ns(|| "non_residue * a1"), &P::NONRESIDUE)?;
         let a0_plus_non_residue_c1 = self
             .c0
             .add(cs.ns(|| "a0 + non_residue * a1"), &non_residue_c1)?;
         let one_plus_non_residue_v0 = v0.mul_by_constant(
-            cs.ns(|| "1 + non_residue * v0"),
+            cs.ns(|| "(1 + non_residue) * v0"),
             &(P::Fp::one() + &P::NONRESIDUE),
         )?;
-
+        // constraint 2
         let c0 = a0_plus_a1
             .mul(
                 cs.ns(|| "(a0 + a1) * (a0 + non_residue * a1)"),
                 &a0_plus_non_residue_c1,
             )?
             .sub(cs.ns(|| "- (1 + non_residue) v0"), &one_plus_non_residue_v0)?;
 
+        // no extra constraint, works directly works on LC:
         v0.double_in_place(cs.ns(|| "2v0"))?;
         let c1 = v0;
 
         Ok(Self::new(c0, c1))
     }
 
+    /* no improvement in number of constraints,
+    improves default implementation by one private variable?
```

Comment:
Probably, we can as well remove this and do like it's done in other functions (i.e. return self.clone().square(other)). Let's mark it as TODO